### ISSUE TITLE
Another issue with caching

### ISSUE THREAD

**Issue by maettu-this** [oleg-shilo/cs-script#99]
Here's another issue with caching. Unfortunately, it's not #86, I have just updated to the latest CSScript source code 3.27.6 to rule that out.

The situation is quite tricky and even trickier to reproduce isolated. I will describe it as detailed as possible, hoping the description gives enough information to give you an idea where the issue could be located. If not sufficient, I could compose a minimal project reproducing the issue. But I suspect, 'minimal' in this case still means scripts + script infos + NUnit based test cases + NUnit based test utilities. So let's first try it the descriptive way:

I have several test scripts for verification of my script hosting application. Instead of running these scripts manually, I use NUnit to run them automatically. The NUnit test cases are also automatically generated based on script info descriptor classes. These classes are decorated with an attribute that is the base for NUnit test generation. Here's an example:
```
//css_import ..\..\..\..\ScriptIDE\ScriptExamples\DataEchoing\Script.cs;

using System.Collections.Generic;

namespace MyProject.Test.Files.DataEchoing
{
	/// <summary>
	/// Information about the script referred to below and imported above.
	/// </summary>
	[MyProject.Test.Utilities.ScriptInfo(                          <= Attribute for test generation.
		@"\DataEchoing\ScriptInfo.cs",                         <= Helper identification.
		typeof(MyProject.Test.Files.DataEchoing.ScriptInfo),   <= Script info lookup, implemented below.
		typeof(MyProject.ScriptExamples.DataEchoingScript),    <= Script lookup, located somewhere else in project.
		@"..\..\..\..\ScriptIDE\ScriptExamples\DataEchoing\Script.cs"
	)]
	public static class MyProject.Test.Files.DataEchoing.ScriptInfo
	{
		/// <summary>
		/// Returns the script's test cases.
		/// </summary>
		public static IEnumerable<MyProject.Test.Utilities.ScriptTestCaseData> TestCases
		{
			get
			{
				yield return (new MyProject.Test.Utilities.ScriptTestCaseData(StandardResult.Success));
			}
		}
	}
}
```
This script info will yield a single test case for the given script. Other infos yield multiple test cases.

When NUnit is started, a `[SetUpFixture]` will collect all `ScriptInfo` attributes, use CSScript to compile and load them...
```
	/// <summary>
	/// Returns all types that define a <see cref="ScriptInfoAttribute"/> from the source code
	/// located at <paramref name="filePath"/>.
	/// </summary>
	private static IEnumerable<Type> GetAllScriptInfoTypes(string filePath)
	{
		var compilerOptions = CompilerOptionsHelper.ComposeDefineImplicitly();
		var assembly = CSScript.LoadWithConfig(filePath, null, false, ScriptHostHelper.StandardSettings, compilerOptions);
		return (assembly.GetTypes().Where(t => t.IsDefined(typeof(ScriptInfoAttribute), true)));
	}
```
...and generate the according test cases which will then run the according test script. Pretty cool. But:

I implemented the test generation using `IEnumerable<>` iterators, same as NUnit does for its test case generation. Accidentally, the iterator calls ended up in executing `CSScript.LoadWithConfig(filePath,...);` twice.

The first test case `DataEchoing` is generated properly, no issue there. But subsequent test cases that also use `ScriptInfo` as test descriptor type name fail. E.g. `DataForwarding`:

The first iterator call seems OK, the given file will be loaded, the assembly states `C:/Users/<User>/AppData/Local/Temp/nunit20/ShadowCopyCache/6708_636523828993833217/Tests_73106732/assembly/dl3/0d5eaa9a/b33991d9_eb94d301/CSScriptLibrary.DLL` as `CodeBase`. `Location` is `""`.

But the second iterator call for this same test case (`DataForwarding`) fails, as the assembly states `C:/Users/<User>/AppData/Local/Temp/CSSCRIPT/Cache/1822138783/ScriptInfo.cs.compiled` as `CodeBase` and `Location`. This looks OK at first sight, but `1822138783` unfortunately contains the script info for the **first** loaded test case (`DataEchoing`)! The correct cache location would be `1982570327`. Same issue with third test case (`FileConversion`) located in `356279381`.

[CSScript Cache.zip](https://github.com/oleg-shilo/cs-script/files/1659223/CSScript.Cache.zip)

Interestingly, there is only an issue with test descriptors with equal name. Test descriptors that have a unique name (e.g. `ScriptInfoOfSimpleExample`) will properly be treated.

I can work around this issue by ensuring the iterator calls only execute `CSScript.LoadWithConfig(filePath,...);` once. That's faster anyway. (I could also work around this issue by explicitly naming each `ScriptInfo` with a unique name. But that just doesn't make sense, they are located in separate namespaces already.) So there is no need to rush with analyzing and fixing this.

**Comment by oleg-shilo** [comment]
I will have a look at your attached zip. Thank you.

A few points that I found peculiar. `Location is ""` means it is an assembly that is not loaded from file but from the memory (in-memory assembly). It is loaded with `Assembly.Load(byte[])` An the assembly in this case is CSScriptLibrary.DLL. This assembly is expected to be ALWAYS loaded from the file. Only scripts can be in-mem assemblies but not the script engine itself. 

NUnit is a tricky environment and it plays with the AppDomains quite intensively. Thus I am not sure that you would have the same behavior if you execute the same testcase in just a ConsoleApp project .

Another concerning point is that "ShadowCopyCache". It sometimes can play nasty tricks. There have already been reports and evidences of it supplying wring assemblies. 

I actually do not expect anything to be loaded from "ShadowCopyCache" in your case but there is that CSScriptLibrary.dll. And you have already  logged #50 and we had a very long exchange about the matter and guess what?  "ShadowCopyCache" was also involved there. I just do not trust this guy :)

**Comment by maettu-this** [comment]
Hmm... NUnit....

A little piece of clarification:

The call `var assembly = CSScript.LoadWithConfig(filePath, null, false, ScriptHostHelper.StandardSettings, compilerOptions);` that results in `Location is ""` does indeed take a script file as `filePath` argument, not a path to a .dll file. Actually, the file is not really a script, just the script info shown further above. But I assume it's OK to compile and load such file, just for the sake of doing some reflection on it.

**Comment by oleg-shilo** [comment]
> The call var assembly = CSScript.LoadWithConfig(filePath, null, false, ScriptHostHelper.StandardSettings, compilerOptions); that results in Location is "" does indeed ...

Something strange. "assembly" is a compiled script assembly file. OK. It may have  `Location:""` if CSSCript.GlobalSettings.InMemoryAssembly = true. Nothing wrong with that. My concern is that your trace had `...eb94d301/CSScriptLibrary.DLL` having `Location:""`. And this is not a script assembly but the script engine assembly.

BTW you can always find the file associated with the script assembly even if `Location:""`. 
```C#
var script = assembly.GetScriptName();

// and the implementation 
public static class CSScriptLibraryExtensionMethods
{
    static public string GetScriptName(this Assembly assembly)
    {
        //Note assembly can contain only single AssemblyDescriptionAttribute
        foreach (AssemblyDescriptionAttribute attribute in assembly.GetCustomAttributes(typeof(AssemblyDescriptionAttribute), true))
            return attribute.Description;
        return null;
    }
```

**Comment by maettu-this** [comment]
`GetScriptName()` indeed returns the given `filePath`, i.e. the piece of code I want to compile and load.
`assembly.ManifestModule.Assembly|ScopeName` also indicate the correct file, i.e. ScriptInfo.dll.
Strange indeed. Probable that it's related to NUnit, as you suspect.

As mentioned, the issue is not time critical at all, so it's fully OK if the issue is pending for a while.

**Comment by oleg-shilo** [comment]
OK then. It'l put it on hold at least for a while. 

BTW, I do remember that NUnit had a check box somewhere in config. Something like "Use shadow copy". I always had to uncheck it.

**Comment by maettu-this** [comment]
Negative, the "Enable Shadow Copy" setting does not have a visible impact, the issue occurs with either setting. Note that I manually deleted the cache before starting NUnit, to rule out any side-effects. The issue remains, though the cache stays empty.

(Also, I somewhat dislike to change something that explicitly states "Warning: ...should normally not be changed" ;-)

**Comment by oleg-shilo** [comment]
:o)

**Comment by maettu-this** [comment]
Hmm, this issue really limits testability of my project. I kind of work around it currently, but the workaround reduces testcoverage. In 2-3 months I expect another major release, I could then spend time trying to condense this issue to the essence, giving you a better chance finding the root cause of it.

**Comment by maettu-this** [comment]
Yepp. I'll let you know as soon I have a reproducible setup. As mentioned, it's gonna take a few months.
